### ISSUE TITLE
Training with multi gpus

### ISSUE THREAD

**Issue by Howardqlz** [facebookresearch/detectron2#3885]
Hey this bug occured when i try to use two gpus training my own dataset , it just stop logging when nearly finish the first inference, no metrics, didn't continue training just stuck. Using one gpu is no problem, and comparing to use 2 gpus, the loss is much lower and better. 
I guess is multi-processes problem, but don't know how to fix it. Wish your help

## Instructions To Reproduce the Issue:
1. I simple rewrite the tools/plain_train_net.py to training my own dataset, most code change in do_train() and setup():
```
logger = logging.getLogger("detectron2")


def get_evaluator(cfg, dataset_name, output_folder=None):
    """
    Create evaluator(s) for a given dataset.
    This uses the special metadata "evaluator_type" associated with each builtin dataset.
    For your own dataset, you can simply create an evaluator manually in your
    script and do not have to worry about the hacky if-else logic here.
    """
    if output_folder is None:
        output_folder = os.path.join(cfg.OUTPUT_DIR, "inference")
    evaluator_list = []
    evaluator_type = MetadataCatalog.get(dataset_name).evaluator_type
    if evaluator_type in ["sem_seg", "coco_panoptic_seg"]:
        evaluator_list.append(
            SemSegEvaluator(
                dataset_name,
                distributed=True,
                output_dir=output_folder,
            )
        )
    if evaluator_type in ["coco", "coco_panoptic_seg"]:
        evaluator_list.append(COCOEvaluator(dataset_name, output_dir=output_folder))
    if evaluator_type == "coco_panoptic_seg":
        evaluator_list.append(COCOPanopticEvaluator(dataset_name, output_folder))
    if evaluator_type == "cityscapes_instance":
        assert (
            torch.cuda.device_count() >= comm.get_rank()
        ), "CityscapesEvaluator currently do not work with multiple machines."
        return CityscapesInstanceEvaluator(dataset_name)
    if evaluator_type == "cityscapes_sem_seg":
        assert (
            torch.cuda.device_count() >= comm.get_rank()
        ), "CityscapesEvaluator currently do not work with multiple machines."
        return CityscapesSemSegEvaluator(dataset_name)
    if evaluator_type == "pascal_voc":
        return PascalVOCDetectionEvaluator(dataset_name)
    if evaluator_type == "lvis":
        return LVISEvaluator(dataset_name, cfg, True, output_folder)
    if len(evaluator_list) == 0:
        raise NotImplementedError(
            "no Evaluator for the dataset {} with the type {}".format(dataset_name, evaluator_type)
        )
    if len(evaluator_list) == 1:
        return evaluator_list[0]
    return DatasetEvaluators(evaluator_list)


def do_test(cfg, model):
    results = OrderedDict()
    for dataset_name in cfg.DATASETS.TEST:
        data_loader = build_detection_test_loader(cfg, dataset_name)
        evaluator = get_evaluator(
            cfg, dataset_name, os.path.join(cfg.OUTPUT_DIR, "inference", dataset_name)
        )
        results_i = inference_on_dataset(model, data_loader, evaluator)
        results[dataset_name] = results_i
        if comm.is_main_process():
            logger.info("Evaluation results for {} in csv format:".format(dataset_name))
            print_csv_format(results_i)
    return results['testSet']


def do_train(cfg, model, best_ap, resume=False):
    model.train()
    optimizer = build_optimizer(cfg, model)
    scheduler = build_lr_scheduler(cfg, optimizer)

    checkpointer = DetectionCheckpointer(
        model, cfg.OUTPUT_DIR, optimizer=optimizer, scheduler=scheduler
    )
    start_iter = (
        checkpointer.resume_or_load(cfg.MODEL.WEIGHTS, resume=resume).get("iteration", -1) + 1
    )
    max_iter = cfg.SOLVER.MAX_ITER

    periodic_checkpointer = PeriodicCheckpointer(
        checkpointer, cfg.SOLVER.CHECKPOINT_PERIOD, max_iter=max_iter
    )

    writers = (
        [
            CommonMetricPrinter(max_iter),
            JSONWriter(os.path.join(cfg.OUTPUT_DIR, "metrics.json")),
            TensorboardXWriter(cfg.OUTPUT_DIR),
        ]
        if comm.is_main_process()
        else []
    )

    # compared to "train_net.py", we do not support accurate timing and
    # precise BN here, because they are not trivial to implement in a small training loop
    data_loader = build_detection_train_loader(cfg)
    logger.info("Starting training from iteration {}".format(start_iter))
    with EventStorage(start_iter) as storage:
        for data, iteration in zip(data_loader, range(start_iter, max_iter)):
            storage.iter = iteration

            loss_dict = model(data)
            losses = sum(loss_dict.values())
            assert torch.isfinite(losses).all(), loss_dict

            loss_dict_reduced = {k: v.item() for k, v in comm.reduce_dict(loss_dict).items()}
            losses_reduced = sum(loss for loss in loss_dict_reduced.values())
            if comm.is_main_process():
                storage.put_scalars(total_loss=losses_reduced, **loss_dict_reduced)

            optimizer.zero_grad()
            losses.backward()
            optimizer.step()
            storage.put_scalar("lr", optimizer.param_groups[0]["lr"], smoothing_hint=False)
            scheduler.step()

            if (
                cfg.TEST.EVAL_PERIOD > 0
                and (iteration + 1) % cfg.TEST.EVAL_PERIOD == 0
                and iteration != max_iter - 1
            ):
                eval_result = do_test(cfg, model)
                cur_ap = float(eval_result["bbox"]["AP"])
                if best_ap < cur_ap:
                    best_ap = cur_ap
                    best_iter = iteration
                    best_result = eval_result["bbox"]
                    checkpointer = DetectionCheckpointer(model, save_dir="./output/new_backbone_best/")
                    checkpointer.save("model_best")
                print("best model at iteration: ", best_iter)
                print("best model AP: ", best_result)
                # Compared to "train_net.py", the test results are not dumped to EventStorage
                comm.synchronize()

            if iteration - start_iter > 5 and (
                (iteration + 1) % 20 == 0 or iteration == max_iter - 1
            ):
                for writer in writers:
                    writer.write()
            periodic_checkpointer.step(iteration)


def setup(args):
    """
    Create configs and perform basic setups.
    """
    cfg = get_cfg()
    trainPath = "./data/new_train/images/"
    trainjsonPath = "./data/new_train/new_train.json"
    testPath = "./data/new_test/images/"
    testjsonPath = "./data/new_test/new_test.json"
    
    cfg.OUTPUT_DIR = './pretrained_model/new_backbone_train_steps'
    cfg.merge_from_file("./detectron2/configs/Misc/cascade_mask_rcnn_R_50_FPN_3x.yaml")
    cfg.MODEL.WEIGHTS = "./pretrained_model/model_final_480dd8.pkl"
    # cfg.merge_from_file("./detectron2/configs/COCO-Detection/faster_rcnn_R_50_FPN_3x.yaml")
    cfg.DATASETS.TRAIN = ("trainSet",)
    cfg.DATASETS.TEST = ("testSet",)
    cfg.SOLVER.MOMENTUM = 0.9
    #cfg.SOLVER.CHECKPOINT_PERIOD = 500
    cfg.SOLVER.IMS_PER_BATCH = 4
    cfg.SOLVER.BASE_LR = 1e-3
    cfg.MODEL.ROI_HEADS.NUM_CLASSES = 43
    cfg.SOLVER.MAX_ITER = 860000
    cfg.TEST.EVAL_PERIOD = 8600
    
    os.makedirs(cfg.OUTPUT_DIR, exist_ok=True)
    
    register_coco_instances("trainSet", {}, trainjsonPath, trainPath)
    register_coco_instances("testSet", {}, testjsonPath, testPath)
    cfg.freeze()
    default_setup(
        cfg, args
    )  # if you don't like any of the default setup, write your own setup code
    return cfg


def main(args):
    cfg = setup(args)
    best_ap = 0
    model = build_model(cfg)
    logger.info("Model:\n{}".format(model))
    if args.eval_only:
        DetectionCheckpointer(model, save_dir=cfg.OUTPUT_DIR).resume_or_load(
            cfg.MODEL.WEIGHTS, resume=args.resume
        )
        return do_test(cfg, model)

    distributed = comm.get_world_size() > 1
    if distributed:
        model = DistributedDataParallel(
            model, device_ids=[comm.get_local_rank()], broadcast_buffers=False
        )

    do_train(cfg, model, best_ap, resume=args.resume)
    return do_test(cfg, model)


if __name__ == "__main__":
    args = default_argument_parser().parse_args()
    print("Command Line Args:", args)
    launch(
        main,
        args.num_gpus,
        num_machines=args.num_machines,
        machine_rank=args.machine_rank,
        dist_url=args.dist_url,
        args=(args,),
    )
```
2. Then run python new_plain_train_net.py --num-gpus 2
3. The logging stop here:
```
...
[01/13 18:14:51 d2.evaluation.evaluator]: Inference done 4299/4313. 0.0726 s / img. ETA=0:00:05
[01/13 18:14:56 d2.evaluation.evaluator]: Inference done 4313/4313. 0.0726 s / img. ETA=0:00:00
[01/13 18:14:56 d2.evaluation.evaluator]: Total inference time: 0:26:54.948302 (0.374872 s / img per device, on 2 devices)
[01/13 18:14:56 d2.evaluation.evaluator]: Total inference pure compute time: 0:05:12 (0.072584 s / img per device, on 2 devices)
```
stuck for long time. If i stop it , some logger like' multi_process timeout() 'will appear.
## Expected behavior:
Compute metrics, continue training

## Environment:
```
----------------------  -----------------------------------------------------------------
sys.platform            linux
Python                  3.6.9 (default, Oct  8 2020, 12:12:24) [GCC 8.4.0]
numpy                   1.16.6
detectron2              0.3 @/home/admin/workspace/dla/Publaynet/detectron2/detectron2
Compiler                GCC 7.5
CUDA compiler           CUDA 10.1
detectron2 arch flags   7.0
DETECTRON2_ENV_MODULE   <not set>
PyTorch                 1.5.0 @/home/admin/.local/lib/python3.6/site-packages/torch
PyTorch debug build     False
GPU available           True
GPU 0,1                 Tesla V100-SXM2-16GB (arch=7.0)
CUDA_HOME               /usr/local/cuda
Pillow                  8.0.1
torchvision             0.6.0 @/home/admin/.local/lib/python3.6/site-packages/torchvision
torchvision arch flags  3.5, 5.0, 6.0, 7.0, 7.5
fvcore                  0.1.5.post20211023
cv2                     4.2.0
----------------------  -----------------------------------------------------------------
PyTorch built with:
  - GCC 7.3
  - C++ Version: 201402
  - Intel(R) Math Kernel Library Version 2019.0.5 Product Build 20190808 for Intel(R) 64 architecture applications
  - Intel(R) MKL-DNN v0.21.1 (Git Hash 7d2fd500bc78936d1d648ca713b901012f470dbc)
  - OpenMP 201511 (a.k.a. OpenMP 4.5)
  - NNPACK is enabled
  - CPU capability usage: AVX2
  - CUDA Runtime 10.2
  - NVCC architecture flags: -gencode;arch=compute_37,code=sm_37;-gencode;arch=compute_50,code=sm_50;-gencode;arch=compute_60,code=sm_60;-gencode;arch=compute_61,code=sm_61;-gencode;arch=compute_70,code=sm_70;-gencode;arch=compute_75,code=sm_75;-gencode;arch=compute_37,code=compute_37
  - CuDNN 7.6.5
  - Magma 2.5.2
```
